hrw.go

```diff
@@ -1,280 +1,123 @@
-	l := len(nodes)
-	sorted := make([]uint64, l)
-	dist := make([]uint64, l)
-	for i := range nodes {
-		sorted[i] = uint64(i)
-		dist[i] = distance(nodes[i], hash)
-	}
-
-	sort.Slice(sorted, func(i, j int) bool {
-		return dist[sorted[i]] < dist[sorted[j]]
-	})
-	return sorted
+type hashableWrapper struct {
+	h uint64
 }
 
-// SortByWeight receive nodes, weights and hash, and sort it by distance * weight.
-func SortByWeight(nodes []uint64, weights []float64, hash uint64) []uint64 {
-	result := make([]uint64, len(nodes))
-	copy(nodes, result)
-	sortByWeight(len(nodes), false, nodes, weights, hash, reflect.Swapper(result))
-	return result
+func (h hashableWrapper) Hash() uint64 {
+	return h.h
 }
 
-// SortSliceByValue received []T and hash to sort by value-distance.
-func SortSliceByValue(slice interface{}, hash uint64) {
-	rule := prepareRule(slice)
-	if rule != nil {
-		swap := reflect.Swapper(slice)
-		sortByDistance(len(rule), false, rule, hash, swap)
-	}
-}
+// Sort defines and sorts the scores for the provided hashable
+// entities against the provided hashable object (in its general
+// sense).
+// See [Hash], [Hashable] and https://en.wikipedia.org/wiki/Rendezvous_hashing.
+func Sort[V, P Hashable](vv []V, object P) {
+	oHash := object.Hash()
 
-// SortSliceByWeightValue received []T, weights and hash to sort by value-distance * weights.
-func SortSliceByWeightValue(slice interface{}, weights []float64, hash uint64) {
-	rule := prepareRule(slice)
-	if rule != nil {
-		swap := reflect.Swapper(slice)
-		sortByWeight(reflect.ValueOf(slice).Len(), false, rule, weights, hash, swap)
-	}
-}
+	var s sliceToSort[V, uint64]
+	s.s = vv
+	s.distances = make([]uint64, len(vv))
 
-// SortSliceByIndex received []T and hash to sort by index-distance.
-func SortSliceByIndex(slice interface{}, hash uint64) {
-	length := reflect.ValueOf(slice).Len()
-	swap := reflect.Swapper(slice)
-	sortByDistance(length, true, nil, hash, swap)
-}
+	for i := range vv {
+		s.distances[i] = distance(vv[i].Hash(), oHash)
+	}
 
-// SortSliceByWeightIndex received []T, weights and hash to sort by index-distance * weights.
-func SortSliceByWeightIndex(slice interface{}, weights []float64, hash uint64) {
-	length := reflect.ValueOf(slice).Len()
-	swap := reflect.Swapper(slice)
-	sortByWeight(length, true, nil, weights, hash, swap)
+	sort.Stable(&s)
 }
 
-func prepareRule(slice interface{}) []uint64 {
-	t := reflect.TypeOf(slice)
-	if t.Kind() != reflect.Slice {
-		panic("HRW sort expects slice, got " + t.Kind().String())
+// SortWeighted is the same as [Sort] but allows using weights for
+// a slice being sorted. A weight is applied to the corresponding
+// element's score in the resulting slice. A weight allows modifying
+// the default probability of an element to win HRW sorting.
+// Value slice's length and weight slice's length MUST be the same.
+// Weights MUST be in [0.0; 1.0] range.
+func SortWeighted[V, P Hashable, W constraints.Float](vv []V, weights []W, object P) {
```

Comment:
Named return variables?

Reply:
it's a generic time 🤩
no named return variables